backends/ebpf/ebpfTable.cpp

```diff
@@ -289,29 +340,75 @@ void EBPFTable::emitInstance(CodeBuilder* builder) {
 }
 
 void EBPFTable::emitKey(CodeBuilder* builder, cstring keyName) {
-    if (keyGenerator == nullptr)
+    if (keyGenerator == nullptr) {
         return;
+    }
+
+    if (isLPMTable()) {
+        builder->emitIndent();
+        builder->appendFormat("%s.%s = sizeof(%s)*8 - %d",
+                              keyName.c_str(), prefixFieldName,
+                              keyName, prefixLenFieldWidth);
+        builder->endOfStatement(true);
+    }
+
     for (auto c : keyGenerator->keyElements) {
         auto ebpfType = ::get(keyTypes, c);
         cstring fieldName = ::get(keyFieldNames, c);
-        CHECK_NULL(fieldName);
+        if (fieldName == nullptr || ebpfType == nullptr)
+            continue;
         bool memcpy = false;
         EBPFScalarType* scalar = nullptr;
-        unsigned width = 0;
+        cstring swap;
         if (ebpfType->is<EBPFScalarType>()) {
             scalar = ebpfType->to<EBPFScalarType>();
-            width = scalar->implementationWidthInBits();
+            unsigned width = scalar->implementationWidthInBits();
             memcpy = !EBPFScalarType::generatesScalar(width);
+
+            if (width <= 8) {
+                swap = "";  // single byte, nothing to swap
+            } else if (width <= 16) {
+                swap = "bpf_htons";
+            } else if (width <= 32) {
+                swap = "bpf_htonl";
+            } else if (width <= 64) {
+                swap = "bpf_htonll";
+            }  // TODO: handle width > 64 bits
+        }
+
+        bool isLPMKeyBigEndian = false;
+        if (isLPMTable()) {
+            if (c->matchType->path->name.name == P4::P4CoreLibrary::instance.lpmMatch.name)
+                isLPMKeyBigEndian = true;
         }
 
         builder->emitIndent();
         if (memcpy) {
-            builder->appendFormat("memcpy(&%s.%s, &", keyName.c_str(), fieldName.c_str());
-            codeGen->visit(c->expression);
-            builder->appendFormat(", %d)", scalar->bytesRequired());
+            if (isLPMKeyBigEndian) {
+                // FIXME: will not work on big endian machines because byte swap
```

Comment:
does this bug exist in ubpf too? I thought you fixed it there. If so, could you backport it?